docs/design/sql/14-automatic-query-resubmission.md

```diff
@@ -0,0 +1,157 @@
+   forwarded to the user
+2. `RETRY_SELECTS`: the query will be retried if:
+   - no rows were received yet
+   - the SQL text starts with `SELECT` (case-insensitive, ignoring
+     white space)
+3. `RETRY_SELECTS_ALLOW_DUPLICATES`: as before, but query can be retried
+   after it returned some rows. The already-returned rows will be returned again.
+4. `RETRY_ALL`: all queries will be retried after a failure. This option is 
+   dangerous as it will retry all queries, including DML, DDL etc. E.g. a retried
+   `CREATE MAPPING` query can fail with "object already exists", if there's no `OR
+   REPLACE` option and the first attempt actually succeeded, so the user will   
+   receive an error even though it was successful, which is very confusing.
+
+The default mode will be `NEVER` (b/w compatible).
+
+#### Ways to pick resubmission policy
+
+1. Ideally, we would be able to configure the retry policy per query, but that's
+not possible. We could add this option to Java API (to the `SqlStatement`
+class), but this is a non-standard API. We expect many clients to use JDBC or
+other standardized APIs in other languages, which don't provide this support.
+
+2. Another idea is that we could specify this as a query hint, e.g. in the form of
+`SELECT /*+ ALLOW_RETRY */`. But for this we would have to parse the query on
+the client, which we want to avoid. Another reason against this is that queries
+are often generated, such as through ORM or other SQL tools, and injecting the
+hint might not be possible or easy.
+
+3. Another option is to use a session parameter, e.g. executing `SET
+RETRY_STATEGY=<mode>`. The client will have to parse these commands, but we will
+not have to provide a full SQL parser.
+
+4. Provide a new option in `ClientConfig`, and a new JDBC URL parameter for 
+JDBC. Other standardized APIs also provide these. The disadvantage is that it
+will apply to all statements executed using that connection, that is also to DDL
+statements etc.
+
+We propose to implement options (3) and (4). Since the option (4) is feasible
+for all modes, except for not-that-useful `RETRY_ALL` mode, we consider the
+option (3) a nice-to-have feature.
+
+#### Explanation of "no rows received yet"
+
+This means that the `SqlResult.iterator()` didn't ever return from `next()`.
+However, there can be an ongoing blocking call to `next()` in another thread.
+
+#### Other configuration options
+
+The Java client retries invocations in this way: first 5 invocations are retried
+without a delay (hardcoded value). Afterwards the delay is `2^<failureCount>`
+ms, up to 1000 ms (configurable by the
+`hazelcast.client.invocation.retry.pause.millis` property). After 120 seconds
+(configurable by the `hazelcast.client.invocation.timeout.seconds` property), a
+failure is reported to the user.
+
+We think this scheme can be also used for SQL retries and there's no need to
+provide more configuration options.
```

Comment:
SQL has a different mechanism to transfer errors to the client. (See `SqlErrorCodec`) If the query throws an SQL error and the member sends a response with it, from the point of view of the invocation system, that invocation succeeded so the retry logic described above will not be used for such error responses. 

The mechanism described above will only apply to SQL when the connection we send the invocation closes etc. Is it enough for you guys? Or will you just use these configuration options and have your own retry logic implemented on top of them?

Reply:
We implemented it also in the handler of SQL errors, we just reused the already-existing configuration options.